File: python/pages/cart_page.py
"""Page object for the shopping cart."""

from __future__ import annotations

from playwright.sync_api import Page, expect

from .base_page import BasePage


class CartPage(BasePage):
    def __init__(self, page: Page) -> None:
        super().__init__(page)
        self.cart_items = page.locator(".cart_item")
        self.checkout_button = page.get_by_test_id("checkout")

    def expect_item_count(self, count: int) -> None:
        expect(self.cart_items).to_have_count(count)

    def remove_item(self, product_name: str) -> None:
        (
            self.cart_items.filter(has_text=product_name)
            .get_by_role("button", name="Remove")
            .click()
        )

    def item_names(self) -> list[str]:
        return self.cart_items.locator(".inventory_item_name").all_inner_texts()

    def checkout(self) -> None:
        self.checkout_button.click()

File: typescript/src/pages/CartPage.ts
import { type Page, type Locator, expect } from '@playwright/test';
import { BasePage } from './BasePage.js';

/**
 * Page object for the shopping cart and the start of the checkout flow.
 */
export class CartPage extends BasePage {
  readonly cartItems: Locator;
  readonly checkoutButton: Locator;
  readonly continueShopping: Locator;

  constructor(page: Page) {
    super(page);
    this.cartItems = page.locator('.cart_item');
    this.checkoutButton = page.getByTestId('checkout');
    this.continueShopping = page.getByTestId('continue-shopping');
  }

  async expectItemCount(count: number): Promise<void> {
    await expect(this.cartItems).toHaveCount(count);
  }

  async removeItem(productName: string): Promise<void> {
    await this.cartItems
      .filter({ hasText: productName })
      .getByRole('button', { name: 'Remove' })
      .click();
  }

  async itemNames(): Promise<string[]> {
    return this.cartItems.locator('.inventory_item_name').allInnerTexts();
  }

  async checkout(): Promise<void> {
    await this.checkoutButton.click();
  }
}
